src/main/scala/scala/compat/java8/converterImpl/MakesSteppers.scala

```diff
@@ -5,27 +5,117 @@ import language.implicitConversions
 import scala.compat.java8.collectionImpl._
 import scala.compat.java8.runtime._
 
-/** Classes or objects implementing this trait create streams suitable for sequential use */
-trait MakesSequentialStream[A, SS <: java.util.stream.BaseStream[A, SS]] extends Any {
-  def seqStream: SS
+trait MakesStepper[T, +Extra] extends Any {
+  /** Generates a fresh stepper of type `S` for element type `T` */
+  def stepper[S <: Stepper[_]](implicit ss: StepperShape[T, S]): S with Extra
 }
 
-/** Classes or objects implementing this trait create streams suitable for parallel use */
-trait MakesParallelStream[A, SS <: java.util.stream.BaseStream[A, SS]] extends Any {
-  def parStream: SS
+trait MakesKeyValueStepper[K, V, +Extra] extends Any {
+  /** Generates a fresh stepper of type `S` over map keys of type `K` */
+  def keyStepper[S <: Stepper[_]](implicit ss: StepperShape[K, S]): S with Extra
+
+  /** Generates a fresh stepper of type `S` over map values of type `V` */
+  def valueStepper[S <: Stepper[_]](implicit ss: StepperShape[V, S]): S with Extra
+}
+
+sealed trait StepperShape[T, S <: Stepper[_]] { def ref: Boolean }
+object StepperShape extends StepperShapeLowPrio {
+  private[this] def valueShape[T, S <: Stepper[_]]: StepperShape[T, S] = new StepperShape[T, S] { def ref = false }
+
+  // primitive
+  implicit val IntValue     = valueShape[Int, IntStepper]
+  implicit val LongValue    = valueShape[Long, LongStepper]
+  implicit val DoubleValue  = valueShape[Double, DoubleStepper]
+
+  // widening
+  implicit val ByteValue    = valueShape[Byte, IntStepper]
+  implicit val ShortValue   = valueShape[Short, IntStepper]
+  implicit val CharValue    = valueShape[Char, IntStepper]
+  implicit val FloatValue   = valueShape[Float, DoubleStepper]
+}
+trait StepperShapeLowPrio {
+  // reference
+  implicit def anyStepperShape[T]: StepperShape[T, AnyStepper[T]] = new StepperShape[T, AnyStepper[T]] { def ref = true }
 }
 
-trait MakesStepper[+T <: Stepper[_]] extends Any {
-  /** Generates a fresh stepper of type `T` */
-  def stepper: T
+/** Superclass for `MakesStepper` implementations which support parallelization. At least the `AnyStepper` case must be
+  * implemented, all others default to building an `AnyStepper` and putting an unboxing conversion on top. */
+trait MakesParStepper[T] extends Any with MakesStepper[T, EfficientSubstep] {
+  def stepper[S <: Stepper[_]](implicit ss: StepperShape[T, S]) = (ss match {
+    case StepperShape.IntValue    => new Stepper.UnboxingIntStepper   (stepper(StepperShape.anyStepperShape[T]).asInstanceOf[AnyStepper[Int   ]]) with EfficientSubstep
```

Comment:
This can't be in StepperShape because all implementations are slightly different (calling `stepper`, `keyStepper` or `valueStepper`, creating a wrapper with out without `EfficientSubstep`). We could potentially improve performance by dispatching on an Int value representing the StepperShape instead of using the object identities (like `StepperShape.IntValue`) so that a TABLESWITCH can be used.